pkg/storage/image.go

```diff
@@ -176,145 +178,103 @@ func (svc *imageService) makeRepoDigests(knownRepoDigests, tags []string, imageI
-				}
-				user = imageConfig.Config.User
-				cacheItem := imageCacheItem{
-					user:         user,
-					size:         size,
-					configDigest: configDigest,
-				}
-				svc.imageCache[image.ID] = cacheItem
+			results, err = svc.appendCachedResult(systemContext, ref, image, results, nil)
+			if err != nil {
+				return nil, err
 			}
-			name, tags, digests := sortNamesByType(image.Names)
-			imageDigest, repoDigests := svc.makeRepoDigests(digests, tags, image.ID)
-			results = append(results, ImageResult{
-				ID:           image.ID,
-				Name:         name,
-				RepoTags:     tags,
-				RepoDigests:  repoDigests,
-				Size:         size,
-				Digest:       imageDigest,
-				ConfigDigest: configDigest,
-				User:         user,
-			})
+		} else {
+			return nil, err
 		}
 	} else {
 		images, err := svc.store.Images()
 		if err != nil {
 			return nil, err
 		}
-		visited := make(map[string]struct{})
-		defer func() {
-			// We built a map using IDs of images that we looked
-			// at, so remove any items from the cache that don't
-			// correspond to any of those IDs.
-			removedIDs := make([]string, 0, len(svc.imageCache))
-			for imageID := range svc.imageCache {
-				if _, keep := visited[imageID]; !keep {
-					// We have cached data for an image
-					// with this ID, but it's not in the
-					// list of images now, so the image has
-					// been removed.
-					removedIDs = append(removedIDs, imageID)
-				}
-			}
-			// Handle the removals.
-			svc.imageCacheLock.Lock()
-			for _, removedID := range removedIDs {
-				delete(svc.imageCache, removedID)
-			}
-			svc.imageCacheLock.Unlock()
-		}()
+		var newImageCache imageCache
 		for _, image := range images {
-			visited[image.ID] = struct{}{}
-			var user string
-			var size *uint64
-			var configDigest digest.Digest
-			svc.imageCacheLock.Lock()
-			cacheItem, ok := svc.imageCache[image.ID]
-			svc.imageCacheLock.Unlock()
-			if ok {
-				user, size, configDigest = cacheItem.user, cacheItem.size, cacheItem.configDigest
-			} else {
-				ref, err := istorage.Transport.ParseStoreReference(svc.store, "@"+image.ID)
-				if err != nil {
-					return nil, err
-				}
-				img, err := ref.NewImageSource(systemContext)
-				if err != nil {
-					return nil, err
-				}
-				size = imageSize(img)
-				configDigest, err = imageConfigDigest(img, nil)
-				img.Close()
-				if err != nil {
-					return nil, err
-				}
-				imageFull, err := ref.NewImage(systemContext)
-				if err != nil {
-					return nil, err
-				}
-				defer imageFull.Close()
-
-				imageConfig, err := imageFull.OCIConfig()
-				if err != nil {
-					return nil, err
-				}
-				user = imageConfig.Config.User
-				cacheItem := imageCacheItem{
-					user:         user,
-					size:         size,
-					configDigest: configDigest,
-				}
-				svc.imageCacheLock.Lock()
-				svc.imageCache[image.ID] = cacheItem
-				svc.imageCacheLock.Unlock()
+			ref, err := istorage.Transport.ParseStoreReference(svc.store, "@"+image.ID)
```

Comment:
this call previously was made only on cache miss, now it is done unconditionally.

Reply:
`appendCachedResult()` only needs `ref` on misses, though, so passing `image.ID` down to `appendCachedResult()` and reconstructing `ref` there would avoid the call.  But this also forces the whole cache to be recomputed every time `ListImages()` is called with no filter, which saves us less work than the version before this PR would have.

Reply:
> so passing image.ID down to appendCachedResult() and reconstructing ref there would avoid the call.
Reason I am not passing it down is because ref is constructed differently for filtered and unfiltered case and I didn't ant to have another level of special casing in there. I didn't check, but  from the method name I figured it is doing nothing more than parsing of a string, so I kept like it is now.

>  But this also forces the whole cache to be recomputed every time ListImages() is called with no filter
Why? ref is used, as you said, on cache miss only, why it causes whole cache to be recomputed on every unfiltered call?

Reply:
Ah, I missed that `appendCachedResult` was always reading from `svc.imageCache` and not the passed-in `newImageCache`.

Reply:
Yes, it reads from `imageCache`, but saves to `newImageCache` if it is passed